R/prep_sedt_resource.R

```diff
@@ -0,0 +1,135 @@
+#' - A data frame containing the named columns in `coords`
+#' - A `sf` or `sfc` object that can be converted to a data frame using [sf::st_coordinates()]
+#'
+#' If resource is an `sf` object or a path to a spatial data file with POLYGON,
+#' MULTIPOLYGON, or any non-POINT geometry, `placement` is used to convert the
+#' input into points. [sf::st_centroid()] (if `placement = "centroid"`) or
+#' [sf::st_point_on_surface()] (if `placement = "surface"` which is the default)
+#' option.
+#' @returns A path to a CSV or TSV file (depending on value of `fileext`).
+#' @keywords internal
+#' @export
+#' @importFrom utils download.file write.csv
+prep_sedt_resource <- function(resource,
+                               coords = c("lon", "lat"),
+                               file = NULL,
+                               fileext = "csv",
+                               placement = "surface",
+                               ...,
+                               arg = caller_arg(resource),
+                               call = caller_env()) {
+  check_sedt_resource(resource, coords = coords, arg = arg, call = call)
+
+  # If resource is a URL it must be a downloadable file, a FeatureLayer, or Table
+  if (is_url(resource)) {
+    if (grepl("rest/services", resource)) {
+      # Download FeatureLayer or Table
+      check_installed("arcgislayers", call = call)
+      resource <- arcgislayers::arc_read(resource, ..., crs = 4326)
+    } else {
+      file <- file %||% file.path(tempdir(), basename(resource))
+      utils::download.file(resource, destfile = file, method = "auto")
+      resource <- file
+    }
+  }
+
+  if (is_string(resource) && file.exists(resource)) {
+    # Return an existing delimited file path as is
+    if (is_delim_path(resource)) {
+      return(resource)
+    }
+
+    # If not a delimited file attempt to read other file types with
+    # `sf::read_sf`
+    check_installed("sf", call = call)
+    resource <- sf::read_sf(resource, ...)
+  }
+
+  # Convert sf object to data frame
+  if (inherits_any(resource, c("sf", "sfc"))) {
+    resource <- convert_to_coords(
+      resource,
+      coords = coords,
+      placement = placement,
+      call = call
+    )
+  }
+
+  file <- file %||% tempfile(fileext = fileext)
+
+  if (is_installed("readr")) {
+    readr::write_csv(
+      resource,
+      file = file
+    )
+  } else {
+    utils::write.csv(
+      resource,
+      file = file
+    )
+  }
+
+  file
+}
+
+#' @noRd
+check_sedt_resource <- function(resource,
+                                ...,
+                                coords = c("lon", "lat"),
+                                allow_null = FALSE,
+                                arg = caller_arg(resource),
+                                call = caller_env()) {
+  if (allow_null && is.null(resource)) {
+    return(invisible(NULL))
+  }
+
+  not_message <- "{.obj_type_friendly {resource}}"
```

Comment:
I didn't quite understand this syntax - do you mind explaining how it works?

Reply:
"obj_type_friendly" is part of the inline syntax supported by the cli package: https://cli.r-lib.org/reference/inline-markup.html

Jenny Bryan has been a convincing proponent of the "never nester" approach to code structure (I think [this video](https://www.youtube.com/watch?v=CFRhGnuXG-4) is where I first encountered the term) so setting a default value for `not_message` avoids the need for an `else` in the following conditional.